server/test/lib/device/device.getDeviceFeaturesStates.test.js

```diff
@@ -0,0 +1,168 @@
+const EventEmitter = require('events');
+const { expect, assert } = require('chai');
+const uuid = require('uuid');
+const { fake } = require('sinon');
+const db = require('../../../models');
+const Device = require('../../../lib/device');
+
+const event = new EventEmitter();
+
+const now = new Date('2000-06-15 23:59:00.000');
+const insertStates = async () => {
+  const queryInterface = db.sequelize.getQueryInterface();
+  const deviceFeatureStateToInsert = [];
+  const statesToInsert = 23 * 60 + 59;
+  for (let i = 0; i < statesToInsert; i += 1) {
+    const startAt = new Date(now.getTime() - 24 * 60 * 60 * 1000);
+    const date = new Date(startAt.getTime() + ((24 * 60 * 60 * 1000) / statesToInsert) * i);
+    deviceFeatureStateToInsert.push({
+      id: uuid.v4(),
+      device_feature_id: 'ca91dfdf-55b2-4cf8-a58b-99c0fbf6f5e4',
+      value: i,
+      created_at: date,
+      updated_at: date,
+    });
+  }
+  await queryInterface.bulkInsert('t_device_feature_state', deviceFeatureStateToInsert);
+};
+
+describe('Device.getDeviceFeaturesStates', function Describe() {
+  this.timeout(15000);
+
+  beforeEach(async () => {
+    const queryInterface = db.sequelize.getQueryInterface();
+    await queryInterface.bulkDelete('t_device_feature_state');
+  });
+  it('should return the current states', async () => {
+    await insertStates();
+    const variable = {
+      getValue: fake.resolves(null),
+    };
+    const stateManager = {
+      get: fake.returns({
+        id: 'ca91dfdf-55b2-4cf8-a58b-99c0fbf6f5e4',
+        name: 'my-feature',
+      }),
+    };
+    const dateState = `${now.getUTCFullYear()}-${`0${now.getUTCMonth() + 1}`.slice(-2)}-${`0${now.getUTCDate()}`.slice(
+      -2,
+    )}`;
+    const deviceInstance = new Device(event, {}, stateManager, {}, {}, variable);
+    const states = await deviceInstance.getDeviceFeaturesStates('test-device-feature', {
+      from: new Date(`${dateState} 00:00:00.000`).toISOString(),
+      to: new Date(`${dateState} 23:59:59.999`).toISOString(),
+    });
+    expect(states).to.have.lengthOf(23 * 60 + 59 - 1);
```

Comment:
why -1 ?

Reply:
Thx @Pierre-Gilles,
If I'm counting correctly: state[0] to state[23*60+59=1439] in total, but the base date is `const now = new Date('2000-06-15 23:59:00.000'); `. So the 1st value is created the day before '2000-06-14 23:59:00.000'
I did this on purpose to test the consideration of the previous day. So in the end we only have to recover from the 2nd state if I'm not being stupid, i.e. state[1] to state[1439] => state[0] to state[1438].

I am wrong ?

Reply:
You're not wrong, but it's not clear in the test what you are testing! 

Can you maybe create one event explicitly the day before, and create this series of events during the day explicitely

Also, I'm not sure you need that much events for your test? 

I guess it's because you copied the other tests, and we needed to test the downsampling, and here you are not downsampling, so maybe just inserting 10 events is fine (and will take less time)

Reply:
Hey @Pierre-Gilles,
I reduced the number of states created in the tests and reviewed the test (3h * 60 min). For the rest sorry I'm not sure I understood correctly, you want me to delete one of the following tests:
  - it('Should return the full 24h existing state of the device feature - with queries "from" and "to" in GMT')
  - it('should return states between 00:01 and 01:30 only values, created_at and device_feature_id - with queries "from" and "to" in UTC')
  - it('should return states between 00:10 and 01:10 with a target between 2000-06-15 00:10 and now using take and skip , only values')
  - it('should return error, device feature does not exist')
  - it('should return error, start date missing')

I thought I had to test every possibility at the option level.